Incorporating Vignette "Writing Recovery Methods for New Ordination Classes"

jbgracey6 opened the pull request:
Hi @corybrunson! Here's the vignette incorporated into {ordr.extra}. My first commit on the pull request also includes the updated documentation. I ran `devtools::install(build_vignettes = TRUE)` and was able to open the vignette via `browseVignettes()`, so I think things are working smoothly. Let me know if things work on your end!

corybrunson commented:
Hi @jbgracey6, this looks mostly good. However, i won't be able to evaluate it until a subtle issue is resolved: The vignette uses a local path to locate the data set that is the basis for part of the analysis. This data set needs to be made part of the package and documented. The book on R packages should have a chapter or section on this issue—let me know if you are or become unsure how to proceed!

jbgracey6 commented:
Hi @corybrunson. Sorry this went under my radar! The latest two commits have the HCW data incorporated into {ordr.extra} using `use_this::use_data()`, and the vignette was updated accordingly. However, the documentation remains to be updated since I am a little unclear after reading the section from _R Packages_. Do I manually create a file in ordr.extra/R and use the `@format` and `@source` tags to describe the data? In that case, I mostly understand what to do, but I am unsure of how to summarize a few columns in the HCW data: "redcap_repeat_instance", "hc_prov", and "occ". I don't believe these are explained in the paper, but maybe you recall what they refer to. Thanks!

jbgracey6 commented:
Hi @corybrunson and thanks for the suggestions! I've covered all of them in this latest round of commits. Also, the HCW data is now documented, omits the rednote column, and is accompanied by a `data.r` file with descriptions of each variable. Could you let me know if the descriptions I gave are too lengthy? Mostly, they are just the questions from the data dictionary you provided along with the endpoints of the corresponding Likert scales.

Additionally, I ran `devtools::check()` after all updates were made. No errors! However, I got the following warning which I'm not sure how to interpret:

Warning message:
   In system2("quarto", "-V", stdout = TRUE, env = paste0("TMPDIR=",  :
     running command '"quarto" TMPDIR=C:/Users/jbgra/AppData/Local/Temp/RtmpumxYol/file14087544688a -V' had status 1

Lastly, I got the following note:

checking for unstated dependencies in vignettes ... NOTE
  'library' or 'require' call not declared from: 'GPArotation'

Would I resolve this by adding {GPArotation} to the "Suggests" section of `DESCRIPTION`? Thanks!

corybrunson commented:
Thanks @jbgracey6, i too encounter no errors on check.

I also encounter no warnings. The one you shared is...unfamiliar to me. Did you fix it on your end with the post-comment commit?

Yes to {GPArotation}.

Also, i meant to say earlier that the first substantive code chunk in the vignette includes some stuff that should be moved or explained. First, i think `scaled_iris` should be moved to where it's first used, or else before that in a standalone chunk, with a comment in the text about what purpose it serves. Second, you access the unexported function `as_tbl_ord_default()` from {ordr}, which needs to be explained. (It's a helper function defined to reduce redundancy and expedite debugging.) In the source code, i use `getFromNamespace()` for this, which i believe is the endorsed way; can you do that in the vignette as well? Ditto `factor_coord()`. (Future contributors won't need to do this in the source code, since it's already there, but in the vignette you can say that you're illustrating how it's done.)

jbgracey6 commented:
Hi @corybrunson. In the latest round of commits, I updated the definitions of any recoverers that depended on the existence of a scores object, since we discovered that was causing errors when working with a covariance matrix. I explained that motivation for defining the recoverers without dependency on scores in the vignette. I also implemented the changes you suggested in your last comment, but I removed the definition of `factor_coord()` as it seemed out of place.

Is it okay if I revise the row recoverers for `stats::factanal()` as we have for `psych::fa()`? I currently get an error when knitting the vignette because the dimensions of the two matrices on line 489 don't agree.

corybrunson commented:
Interesting problem. Can you describe what the two `recover_rows()` functions return? Could it be a matter of returning only active elements, or some other specification? Or, if the values are different, could they be resolved by conferring inertia appropriately?

jbgracey6 commented:
Yes! I have defined `fa_stats` and `fa_psych` to be equivalent outputs of `factanal()` and `psych::fa()`, respectively. Running `recover_rows(fa_psych)` returns the 0x2 matrix with the column names of the loadings matrix, which is what I expect based on how we defined the recoverer. But `recover_rows(fa_stats)` returns the 6x2 loadings matrix, which doesn't reflect the definition for `recover_rows.factanal()` (reproducible example below).

``` r
library(ordr)
#> Warning: package 'ordr' was built under R version 4.3.3
#> Loading required package: ggplot2
#> Warning: package 'ggplot2' was built under R version 4.3.3
fa_stats <- factanal(scale(swiss), factors = 2, rotation = "none", scores = "regression")
fa_stats$loadings
#> 
#> Loadings:
#>                  Factor1 Factor2
#> Fertility        -0.674   0.356 
#> Agriculture      -0.648   0.297 
#> Examination       0.713  -0.471 
#> Education         0.997         
#> Catholic         -0.178   0.953 
#> Infant.Mortality -0.104   0.169 
#> 
#>                Factor1 Factor2
#> SS loadings      2.419   1.373
#> Proportion Var   0.403   0.229
#> Cumulative Var   0.403   0.632
recover_rows(fa_stats) # equal to loadings
#>                     Factor1     Factor2
#> Fertility        -0.6735941  0.35581357
#> Agriculture      -0.6483598  0.29654465
#> Examination       0.7127582 -0.47079410
#> Education         0.9971815  0.02517316
#> Catholic         -0.1783582  0.95261607
#> Infant.Mortality -0.1044027  0.16920176
```

<sup>Created on 2025-05-02 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

jbgracey6 commented:
Ah, I'd been installing from CRAN, thanks! 

Having removed the chunk defining `factor_coord()` as it was defined in `ordr/R/utils.r` (commit c525800), I do encounter a new error when knitting the vignette. This error persists when I restart my R session and install `ordr` from GitHub. I could of course add the function definition back into the vignette, but I don't think this should be necessary for the vignette to knit, right? Is there something else I might be missing?

corybrunson commented:
`factor_coord()` is not exported in {ordr}, so it cannot be used without the `:::` accessor—similar to how `as_tbl_ord()` and `ordinate()` are currently accessed, though i believe they should _not_ require the `:::`.

jbgracey6 commented:
Hi @corybrunson! My proof of the conference of inertia and validation of recovery of inertia is now in the vignette! I had no trouble knitting it after re-installing, but let me know if you would like me to send an HTML of it to you. As for the last commit, I was not able to run a `git pull` without first committing a change to the `.Rproj` file which I think happened automatically--or by accident--at some point. This is the whole change:

$ git diff
diff --git a/ordr.extra.Rproj b/ordr.extra.Rproj
index 8de5cfd..3f01a6c 100644
--- a/ordr.extra.Rproj
+++ b/ordr.extra.Rproj

Let me know if it seems unimportant; I just didn't want to push an unrelated commit without explanation. Thanks!

corybrunson commented:
Thanks @jbgracey6. I don't really understand the `ProjectId` field in the '.Rproj' file either; i've been studiously avoiding committing it when it manifests, but it was bound to get caught eventually.

It looks like a merge was not successfully completed on your end, hence the '.orig' file. Please make sure that 'methods-psych-fa.R' contains what you want it to, and that nothing is needed from 'methods-psych-fa.R.orig', then delete the latter file.

jbgracey6 commented:
@corybrunson Ah, okay. The latest commit includes the deletion of that file and an update to the surviving methods script to reflect the conference of inertia we discussed/which is validated in the vignette.

corybrunson commented:
@jbgracey6 nice, thank you.

I'll be preoccupied through Tuesday, but after that we should talk about how to restructure the vignette. For example, we might figure out how to "hide" by default the numerical comparisons you use to verify that two matrices are equal or approximately equal, so that the reader can "expand" them if they wish but more conveniently read the exposition if not. We might also save space by linking to topics that are documented in detail elsewhere in the package or its dependencies ({ordr} and {psych}).

jbgracey6 commented:
@corybrunson The latest commit just contains a new bit in the introduction describing the intended audience for the vignette, which was a suggestion Cameron gave upon her last reading of the vignette.
